llama-index-integrations/llms/llama-index-llms-llamafile/llama_index/llms/llamafile/base.py

```diff
@@ -0,0 +1,276 @@
+import io
+import json
+from typing import Any, Dict, Sequence, Tuple
+
+import httpx
+from httpx import Timeout
+
+from llama_index.core.base.llms.types import (
+    ChatMessage,
+    ChatResponse,
+    CompletionResponse,
+    CompletionResponseGen,
+    LLMMetadata,
+    MessageRole,
+)
+from llama_index.core.bridge.pydantic import Field
+from llama_index.core.llms.callbacks import llm_chat_callback, llm_completion_callback
+from llama_index.core.llms.custom import CustomLLM
+
+DEFAULT_REQUEST_TIMEOUT = 30.0
+
+
+def get_addtional_kwargs(
+    response: Dict[str, Any], exclude: Tuple[str, ...]
+) -> Dict[str, Any]:
+    return {k: v for k, v in response.items() if k not in exclude}
+
+
+class Llamafile(CustomLLM):
+    """llamafile lets you distribute and run large language models with a
+    single file.
+
+    To get started, see: https://github.com/Mozilla-Ocho/llamafile
+
+    To use this class, you will need to first:
+
+    1. Download a llamafile.
+    2. Make the downloaded file executable: `chmod +x path/to/model.llamafile`
+    3. Start the llamafile in server mode:
+
+        `./path/to/model.llamafile --server --nobrowser`
+    """
+
+    base_url: str = Field(
+        default="http://localhost:8080",
+        description="Base url where the llamafile server is listening.",
+    )
+
+    request_timeout: float = Field(
+        default=DEFAULT_REQUEST_TIMEOUT,
+        description="The timeout for making http request to llamafile API server",
+    )
+
+    #
+    # Generation options
+    #
+    temperature: float = Field(
+        default=0.8,
+        description="The temperature to use for sampling.",
+        gte=0.0,
+        lte=1.0,
+    )
+
+    seed: int = Field(default=0, description="Random seed")
+
+    additional_kwargs: Dict[str, Any] = Field(
+        default_factory=dict,
+        description="Additional options to pass in requests to the llamafile API.",
+    )
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "llamafile_llm"
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        """LLM metadata."""
+        return LLMMetadata(
+            is_chat_model=True,  # llamafile has OpenAI-compatible chat API for all models
```

Comment:
If llama file is openai compatible, I wonder if it's possible to just extend the `OpenAILike` class here?

Reply:
Was not aware of that class, I was just following the example of other integration classes like Ollama. 

The problem with extending `OpenAILike` here is that only the llamafile chat API `/v1/chat/completions` is OpenAI-compatible. The API calls involved in the `complete` method raise errors. 

I could override the `complete` methods but class will look like 50% the same as it is now and, when those methods get called, a lot of the internal plumbing (like the OpenAI client) just won't get used at all. 

Happy to go ahead and make this change if you want though

Reply:
Ah yea that's fair. No worries in that case